src/obc/providers/obf.py

```diff
@@ -0,0 +1,463 @@
+
+        if "access_token" not in json_response:
+            raise AuthenticationError(
+                (
+                    "Cannot get an access token from the OBF server with provided "
+                    "credentials"
+                )
+            )
+
+        return json_response.get("access_token")
+
+    @property
+    def _get_auth(self):
+        """Make access token generation dynamic."""
+        return OAuth2AccessToken(
+            self._access_token(
+                self.client_id,
+                self.client_secret,
+                self.api_version_prefix,
+                self.api_root_url,
+            )
+        )
+
+    def check_auth(self):
+        """Check OBF API credentials using the dedicated endpoint."""
+        url = f"/ping/{self.client_id}"
+        response = self.get(url)
+        if response.status_code != requests.codes.ok:  # pylint: disable=no-member
+            raise AuthenticationError("Invalid access token for OBF")
+        return response
+
+    @classmethod
+    def iter_json(cls, response: requests.Response) -> Iterable:
+        """Iterate over JSON lines serialization in API responses.
+
+        When multiple objects are returned by the API, they are streamed as
+        JSON lines instead of a JSON list, leading the response.json() method
+        to fail as it is expected a valid JSON list instead. We mitigate this
+        issue by forcing JSON serialization of each item in the response.
+        """
+        try:
+            json_response = response.json()
+            if isinstance(json_response, list):
+                yield from json_response
+            else:
+                yield json_response
+        except requests.JSONDecodeError:
+            for line in response.iter_lines():
+                yield json.loads(line)
+
+    # pylint: disable=arguments-differ
+    def request(self, method, url, **kwargs):
+        """Make OBF API usage more developer-friendly.
+
+        - Automatically add the API root URL so that we can focus on the endpoints
+        - Automatically renew access token when expired
+        """
+        url = urljoin(self.api_root_url, f"{self.api_version_prefix}/{url}")
+        response = super().request(method, url, **kwargs)
+
+        # Try to regenerate the access token in case of 403 response
+        if (
+            response.status_code
+            == requests.codes.forbidden  # pylint: disable=no-member
+        ):
+            # Clear cached property and force access token update
+            self._access_token.cache_clear()
+            self.auth = self._get_auth
+            # Give it another try
+            return super().request(method, url, **kwargs)
+
+        # Check response status and raise an exception on error
+        if self.raise_for_status:
+            response.raise_for_status()
+
+        return response
+
+
+class Badge(BaseModel):
+    """Open Badge Factory Badge Model."""
+
+    id: NoneStr = None
```

Comment:
Is this type enough for the id ?

Reply:
Yes, OBF badge identifier are not "standards", _e.g._ `RPG8PZa4J4aDD`. Not sure we want to be pedantic on this.